content/post/2025-02-01-lazy/index.Rmd

```diff
@@ -0,0 +1,225 @@
+
+Slightly tweaked from dtplyr README,
+
+```{r}
+# load packages
+library(data.table)
+library(dtplyr)
+library(dplyr, warn.conflicts = FALSE)
+
+# create a “lazy” data table that tracks the operations performed on it.
+mtcars2 <- lazy_dt(mtcars)
+
+# create the query
+summary <- mtcars2 %>% 
+  filter(wt < 5) %>% 
+  mutate(l100k = 235.21 / mpg) %>% # liters / 100 km
+  group_by(cyl) %>% 
+  summarise(l100k = mean(l100k))
+
+# the object is lazy, the value is not computed yet
+summary
+nrow(summary)
+
+# we explictly request the data, so now it's there
+answer <- as_tibble(summary)
+nrow(answer)
+
+```
+
+-   the [duckplyr package](https://duckplyr.tidyverse.org/dev/), which is a drop-in replacement for dplyr, powered by DuckDB for fast operation. Objects created with `duckdb_tibble(.lazy = TRUE)` (`.lazy = TRUE` is not the default) and with `read_parquet_duckdb()` (`lazy=TRUE` is the default) are lazy, the value is only computed if explicitly accessed.
+
+### duckplyr, lazy evaluation and deferred evaluation
+
+Some doubts to be clarified
+
+## Lazy as in lazy loading of data in packages (`LazyData`)
+
+If your R package exports data, and sets the `LazyData` field in `DESCRIPTION` to `true`, then the exported datasets are lazily loaded: they're available without the use of `data()`, but they're not actually taking up memory until they are accessed.
+
+There's more details on `LazyData` in the [R packages book by Hadley Wickham and Jenny Bryan](https://r-pkgs.org/data.html#sec-data-data) and in [Writing R Extensions](https://cloud.r-project.org/doc/manuals/r-devel/R-exts.html#Data-in-packages).
+
+Note that internal data is always lazily loaded, and that data that is too big cannot be lazily loaded.
+
+## Lazy as in frugal file modifications
+
+The `pkgdown::build_site()` function, that creates a documentation website for an R package, features a [`lazy` argument](https://pkgdown.r-lib.org/reference/build_site.html#arg-lazy).
+"If `TRUE`, will only rebuild articles and reference pages if the source is newer than the destination."
+
+It is a much simpler concept of laziness: decide right now whether it is needed to rebuild each page.
+
+The potools package, that provides tools for portability and internationalization of R packages, uses ["lazy" for a similar meaning](https://michaelchirico.github.io/potools/reference/po_update.html?q=lazy#ref-usage).
+
+## Lazy as in lazy quantifiers in regular expressions
+
+In regular expressions you can use [quantifiers](https://blog.djnavarro.net/posts/2024-12-16_regex-backreferences/#quantifiers) to indicate how many times a pattern must appear: the pattern can be optional, appear several times, etc.
+You can also specify whether the tool should match as many repetitions as possible, or the fewest number of repetitions possible.
+
+Matching the fewest number of repetitions possible is "lazy" (stingy, I made that up).
+Matching as many repetitions as possible is "eager" (or greedy, actual term).
```

Comment:
Without knowing what the official terms are, this might be slightly confusing? Also, I do see search results with both  lazy and stingy?

Reply:
and I thought I had made that up  ahahah thanks for checking